src/history/hs.c

```diff
@@ -971,11 +989,82 @@ __wt_find_hs_upd(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE **upd
 }
 
 /*
- * __hs_delete_key --
+ * __wt_hs_delete_key --
  *     Delete an entire key's worth of data in the history store.
  */
+int
+__wt_hs_delete_key(WT_SESSION_IMPL *session, uint32_t btree_id, const WT_ITEM *key)
+{
+    WT_CURSOR *hs_cursor;
+    WT_DECL_RET;
+    WT_ITEM hs_key;
+    wt_timestamp_t hs_start_ts;
+    uint64_t hs_counter;
+    uint32_t hs_btree_id, session_flags;
+    int cmp, exact;
+    bool is_owner;
+
+    session_flags = session->flags;
+
+    /*
+     * Some code paths such as schema removal involve deleting keys in metadata and assert that we
+     * shouldn't be opening new dhandles. We won't ever need to blow away history store content in
+     * these cases so let's just return early here.
+     */
+    if (F_ISSET(session, WT_SESSION_NO_DATA_HANDLES))
+        return (0);
+
+    WT_RET(__wt_hs_cursor(session, &session_flags, &is_owner));
+    hs_cursor = session->hs_cursor;
+    /*
+     * In order to delete a key range, we need to be able to inspect all history store records
+     * regardless of their stop time pairs.
+     */
+    F_SET(session, WT_SESSION_IGNORE_HS_TOMBSTONE);
+retry:
```

Comment:
The number of `goto` statements here isn't ideal. Maybe we can talk about how to structure the code differently?